pkg/reconciler/taskrun/taskrun.go

```diff
@@ -1056,3 +1061,25 @@ func retryTaskRun(tr *v1.TaskRun, message string) {
 	taskRunCondSet := apis.NewBatchConditionSet()
 	taskRunCondSet.Manage(&tr.Status).MarkUnknown(apis.ConditionSucceeded, v1.TaskRunReasonToBeRetried.String(), message)
 }
+
+func DeriveTimeoutForRequeueWaitTime(elapsed time.Duration) time.Duration {
+	timeout := time.Duration(math.MaxInt64)
+	// an explicit choice has been made to not make these demarcation times configurable
+	// wrt wait times when timeout has been disabled; rationale:  allowing
+	// disablement of timeout is problematic, but the feature has already been
+	// exposed.  We'll get it to behave (i.e. no negative wait times) but we
+	// will not add configuration to encourage use of the feature.
+	switch {
+	case elapsed < 30*time.Minute:
+		timeout = 30 * time.Minute
+	case elapsed < time.Hour:
+		timeout = time.Hour
+	case elapsed < 6*time.Hour:
+		timeout = 6 * time.Hour
+	case elapsed < 24*time.Hour:
+		timeout = 24 * time.Hour
+	case elapsed < 48*time.Hour:
```

Comment:
We need to document this behavior.

Reply:
I tend to agree with @khrm, we need to document these times somehow.

Reply:
Any suggestions as to where to document @khrm @vdemeester ?  The timeout feature is certainly discussed in the pipelinerun and taskrun API markdown docs.  But I'm not sure if we should clutter API level discussion with some lower level controller implementation detail.

And there is not a current "controller implementation details" markdown.  And I'm not sure if this warrants the creation of such a thing.

And do we drop some hint of the pros and cons of disabling timeouts?

Pending responses from you all, in the interim, I'll work on a doc update to the pipelinerun/taskrun api doc, where timeout is discussed.  For the moment at least, I'll refrain from any disabling timeouts warnings.  Perhaps such warning necessitate broader community agreement.

Thanks.

Reply:
I believe I need to expand the 0 timeout checks to the other finally related timeouts, make sure there are not negative wait times for those as well

Reply:
@gabemontero If I understand correctly, as it is an internal / implementation detail thing, and very similar to something like the "global resync period" of the controller, which is not documented anywhere, might not make sense to document it ; at least in API / users docs. I agree with that, so maybe just some comments would make sense initially 🙃. I wish we had more "deep technical doc", but as we don't have them today, I don't want to block the PR on this indeed.

Reply:
I think you have captured it correctly @vdemeester 

That said, any thoughts from anyone on my simplification for the what wait time to use that I noted in https://github.com/tektoncd/pipeline/pull/7589#discussion_r1463946082 ?

To some degree these staggered wait times I essentially just grabbed from the air doesn't feel ideal.

Also, with these requeues, if the item requeued with wait gets updated, it get reconciled on the update anyway, correct?  So just using a constant value perhaps is just simpler / better, no?

Reply:
if my explanation is not clear, I can submit the alternative as a separate commit so folks can compare